src/nginx/t/auth_asymmetrickey.t

```diff
@@ -0,0 +1,311 @@
+  server {
+    listen 127.0.0.1:${NginxPort};
+    server_name localhost;
+    location / {
+      endpoints {
+        api service.pb.txt;
+        %%TEST_CONFIG%%
+        on;
+      }
+      proxy_pass http://127.0.0.1:${BackendPort};
+    }
+  }
+}
+EOF
+
+my $es256_pubkey = <<'EOF';
+{
+ "keys": [
+  {
+   "kty": "EC",
+   "crv": "P-256",
+   "x": "lqldKduURoauGtQskOXRTTociai06C-Ug_lwDqcXdd4",
+   "y": "t3FPM5-BhLsjyTG6QcDkTotU6PTMmrT6KCfr4L_0Lhk",
+   "alg": "ES256",
+   "kid": "1a"
+  }
+ ]
+}
+EOF
+
+my $rs256_pubkey = <<'EOF';
+{
+ "keys": [
+  {
+   "kty": "RSA",
+   "n": "zaS0LKbCovc6gdmwwEbovLBqEuat2ihKmuXMEAh7yjk--Pw55djgkpiAFaoTr0-iEnJB8QKQAkssU5mQcKHCtKRfVH9TZv3JC8mXeSg1dvS-AckkGqXwuPpYyaTUDZsd7u3xW3lSX4QtrLNcwCo0TRFmUGcpkecy6omJdD8kwhWXYOEkDPZqZXlvWkLfyuelWE8Wcrv-X_v8UrCMOOECRPRxl5tmC93vMnZZAHN35gyLizaPOkXPR69DN-_d34aiLctphiqzTJUlMlpIU2SciXj2CaOMFzioy-cRb9sbr8eN91cDPDs4r-EiFB6bcoAJxaHCyxdhJYihFGfwGjhCkQ",
+   "e": "AQAB",
+   "alg": "RS256",
+   "kid": "2b"
+  }
+ ]
+}
+EOF
+
+my $es256_token = "eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjFhIn0.".
+"eyJpc3MiOiJ0ZXN0LWVzcC1hdXRoLmNvbSIsInN1YiI6InRlc3QtZXNwLWF1dGguY29tIi".
+"wiYXVkIjoib2tfYXVkaWVuY2VfMSJ9.BUmszufjBD1ID2BBvcFQNiXwhSfhfoLuFhO2e0i".
+"aPashTZCmcSn98lFGic2uFMlAzO5rdvF4SQTirX3vpp4spA";
+
+my $rs256_token = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjJiIn0".
+".eyJpc3MiOiJ0ZXN0LWVzcC1hdXRoLmNvbSIsInN1YiI6InRlc3QtZXNwLWF1dGguY29t".
+"IiwiYXVkIjoib2tfYXVkaWVuY2VfMSJ9.KMBjBmA5h1gLspYsC5HnHzyHgmeGPVGeNX0i".
+"ZDY4V5v8GUvBCZEU_SGBIostrhmAfvV1TY4KOuvDT6CcsbDjfm-04_66AhmVOg72y6HxM".
+"E3hhLgNt_pu1Rrn2gi5RMEVMVw50ogT4XUP3CxmLgqwcbesWYmSeStTXx-U556qO_j0qF".
+"7g6OPmhUbmaj9oewCcqI3BHRiCju-WTUKu2qVhQzVtgXSoS2svCgEpawHQ1C6lM51EHoC".
+"ETCV7pd5f2I6GWoCWV3PZ3zM3k2h5EwJs-95oJxWuTcwmOAXABd1h3ySjzD697BNU5sAK".
+"ygtL7KbWkaO6cL0zTOuPIopbPGI91Q";
+
+################################################################################
+# No JWT passed in.
+
+$t->run_daemon(\&bookstore, $t, $BackendPort, 'bookstore.log');
+$t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&key, $t, $PubkeyPort, $es256_pubkey, 'key.log');
+
+is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'Bookstore socket ready.');
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${PubkeyPort}"), 1, 'Pubkey socket ready.');
+
+$t->run();
+
+my $response = ApiManager::http($NginxPort,<<"EOF");
+GET /shelves?key=this-is-an-api-key HTTP/1.0
+Host: localhost
+
+EOF
+
+like($response, qr/HTTP\/1\.1 401 Unauthorized/, 'Returned HTTP 401, missing creds.');
+like($response, qr/WWW-Authenticate: Bearer/, 'Returned auth challenge.');
+like($response, qr/Content-Type: application\/json/i,
+     'Missing creds returned application/json body.');
+like($response, qr/JWT validation failed: Missing or invalid credentials/i,
+     "Error body contains 'Missing or invalid credentials'.");
+
+$t->stop_daemons();
```

Comment:
why do you need to stop and re-start again,  can we skip line 162 to 175?

Reply:
I was thinking to avoid the effect of auth cache. But Since the first three test cases out of four are not cache hit, it would be enough to cover es/rs parse and verify. So I agree with not stop and restart for every test case.